### PR TITLE
Add support for Netpbm .pnm (.ppm/.pgm/.pbm) images using OpenCV

### PR DESCRIPTION
**Pull request by matthew-frank**
Netpbm .pnm files (also known as .ppm, .pgm, or .pbm) is a widely supported uncompressed raw format.  The format is fully supported by OpenCV's imread/imwrite: https://docs.opencv.org/3.0-beta/modules/imgcodecs/doc/reading_and_writing_images.html#imread.

### CODE REVIEW

File: dali/image/image_factory.cc [hunk 1]
```diff
@@ -73,15 +78,17 @@ bool CheckIsTiff(const uint8_t *tiff, int size) {
 std::unique_ptr<Image>
 ImageFactory::CreateImage(const uint8_t *encoded_image, size_t length, DALIImageType image_type) {
   DALI_ENFORCE(CheckIsPNG(encoded_image, length) + CheckIsBMP(encoded_image, length) +
-               CheckIsGIF(encoded_image, length) + CheckIsJPEG(encoded_image, length)
-             + CheckIsTiff(encoded_image, length) == 1,
+               CheckIsGIF(encoded_image, length) + CheckIsJPEG(encoded_image, length) +
+               CheckIsPNM(encoded_image, length) == 1,
```
Comment (JanuszL) [inline dali/image/image_factory.cc]: I think you have removed Tiff by accident

Comment (matthew-frank) [reply dali/image/image_factory.cc]: You are correct.  I will fix.

**Comment by matthew-frank** [comment]
Honestly, this code is completely untested.  I don't even know how to build the DALI codebase, so I'm not even positive it all compiles.

**Comment by JanuszL** [comment]
> Honestly, this code is completely untested. I don't even know how to build the DALI codebase, so I'm not even positive it all compiles.

We will check that.

**Comment by JanuszL** [comment]
Builds 662282

File: dali/image/pnm.cc [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+        GenericImage(encoded_buffer, length, image_type) {
+}
+
+
+Image::ImageDims PnmImage::PeekDims(const uint8_t *pnm, size_t length) const {
+  DALI_ENFORCE(pnm);
+
+  // http://netpbm.sourceforge.net/doc/ppm.html
+  uint8_t* end_ptr = pnm + length;
+  uint8_t* at_ptr  = pnm + 1;
```
[the remark this answers is not in the document]
Comment (JanuszL) [reply dali/image/pnm.cc]: Fixed

File: dali/image/pnm.cc [hunk 3]
```diff
@@ -0,0 +1,72 @@
+// Copyright (c) 2017-2018, NVIDIA CORPORATION. All rights reserved.
```
[the remark this answers is not in the document]
Comment (JanuszL) [reply dali/image/pnm.cc]: Fixed

File: dali/image/pnm.h [hunk 4]
```diff
@@ -0,0 +1,35 @@
+// Copyright (c) 2017-2018, NVIDIA CORPORATION. All rights reserved.
```
[the remark this answers is not in the document]
Comment (JanuszL) [reply dali/image/pnm.h]: Fixed

**Comment by JanuszL** [comment]
Rebuilds 663331

**Comment by JanuszL** [comment]
Builds #698170

**Comment by JanuszL** [comment]
#699097

**Comment by JanuszL** [comment]
#699145

**Comment by JanuszL** [comment]
#699358

**Comment by Kh4L** [comment]
Build 699547

File: dali/image/pnm.cc [hunk 5]
```diff
@@ -50,20 +50,18 @@ Image::ImageDims PnmImage::PeekDims(const uint8_t *pnm, size_t length) const {
...
           } while (!isspace(*at_ptr));
-          continue;
       } else if (isspace(*at_ptr)) {
           if (++state < STATE_DONE) {
               do {
                   ++at_ptr;
                   DALI_ENFORCE(at_ptr < end_ptr);
               } while (isspace(*at_ptr));
-              continue;
           }
```
Comment (JanuszL) [inline dali/image/pnm.cc]: Without this continue you will eat first, "non space" character.

**Pull request merged**